File: app/src/main/java/com/example/routecraft/features/main/RouteAdapter.java
package com.example.routecraft.features.main;

import android.graphics.Color;
import android.graphics.Typeface;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.PopupMenu;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.example.routecraft.R;
import com.example.routecraft.data.pojos.Route;

import java.util.ArrayList;
import java.util.List;

public class RouteAdapter extends RecyclerView.Adapter<RouteAdapter.CustomViewHolder>{

    private List<Route> routeList = new ArrayList<>();
    private Listener listener;

    public RouteAdapter(Listener listener) {
        this.listener = listener;
    }

    public interface Listener {
        void routeClicked(Route route);
        void openRenameRouteDialog(Route route);
        void openDeleteRouteDialog(Route route);
    }

    public void setRouteList(List<Route> routeList){
        this.routeList = routeList;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_route, parent, false);
        return new CustomViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder holder, int position) {
        holder.routeName.setText(routeList.get(position).getRouteName());

//        if(routeList.get(position).isSelected()){
//            holder.container.setBackgroundColor(Color.parseColor("#e0e0e0"));
//            holder.routeName.setTextColor(Color.parseColor("#167ef9"));
//            holder.routeName.setTypeface(Typeface.DEFAULT_BOLD);
//            holder.routeMenuIb.setBackgroundColor(Color.parseColor("#e0e0e0"));
//        }else{
//            holder.container.setBackgroundColor(Color.WHITE);
//            holder.routeName.setTextColor(Color.parseColor("#515767"));
//            holder.routeName.setTypeface(Typeface.DEFAULT);
//            holder.routeMenuIb.setBackgroundColor(Color.WHITE);
//        }

    }

    @Override
    public int getItemCount() {
        return routeList.size();
    }

    class CustomViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, PopupMenu.OnMenuItemClickListener {

        private ConstraintLayout root;
        private ConstraintLayout container;
        private TextView routeName;
        private ImageButton routeMenuIb;

        CustomViewHolder(View itemView) {
            super(itemView);
            root = itemView.findViewById(R.id.root);
            root.setOnClickListener(this);
            container = itemView.findViewById(R.id.container);
            routeMenuIb = itemView.findViewById(R.id.route_menu_ib);
            routeMenuIb.setOnClickListener(this);
            routeName = itemView.findViewById(R.id.route_name_tv);
        }

        public void onClick(View v) {

            if(v == root){
                listener.routeClicked(routeList.get(getAdapterPosition()));
            }else if(v == routeMenuIb) {
                PopupMenu popupMenu = new PopupMenu(v.getContext(), v);
                popupMenu.setGravity(Gravity.END);
                popupMenu.inflate(R.menu.navigation_view_route_menu);
                popupMenu.setOnMenuItemClickListener(this);
                popupMenu.show();
            }
        }

        @Override
        public boolean onMenuItemClick(MenuItem item) {
            switch (item.getItemId()){
                case R.id.action_popup_rename:
                    listener.openRenameRouteDialog(routeList.get(getAdapterPosition()));
                    return true;
                case R.id.action_popup_delete:
                    listener.openDeleteRouteDialog(routeList.get(getAdapterPosition()));
                    return true;
                default:
                    return false;
            }
        }
    }
}

File: app/src/main/java/com/example/routecraft/features/main/RouteAdapter.kt
package com.example.routecraft.features.main

import android.graphics.Color
import android.graphics.Typeface
import android.view.*
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.routecraft.R
import com.example.routecraft.data.pojos.Route
import com.example.routecraft.databinding.ItemRouteBinding

class RouteAdapter(private val listener: Listener) : ListAdapter<Route, RouteAdapter.RouteViewHolder>(DiffCallback()) {

    interface Listener {
        fun routeClicked(route: Route)
        fun renameRouteBtnClicked(route: Route)
        fun deleteRouteBtnClicked(route: Route)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RouteViewHolder {
        val binding = ItemRouteBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return RouteViewHolder(binding, listener)
    }

    override fun onBindViewHolder(holder: RouteViewHolder, position: Int) {
        val currentRoute = getItem(position)
        holder.bind(currentRoute)
    }

    inner class RouteViewHolder(private val binding: ItemRouteBinding, private val listener: Listener) : RecyclerView.ViewHolder(binding.root){

        init {
            binding.apply {
                container.setOnClickListener {
                    val position = adapterPosition
                    if(position != RecyclerView.NO_POSITION){
                        listener.routeClicked(getItem(position))
                    }
                }
                routeMenuIb.setOnClickListener{
                    val position = adapterPosition
                    if(position != RecyclerView.NO_POSITION){
                        val popupMenu = PopupMenu(it.context, it)
                        popupMenu.gravity = Gravity.END
                        popupMenu.inflate(R.menu.menu_navigation_view_route)
                        popupMenu.setOnMenuItemClickListener {item ->
                            when (item.itemId) {
                                R.id.action_popup_rename -> {
                                    listener.renameRouteBtnClicked(getItem(position))
                                    true
                                }
                                R.id.action_popup_delete -> {
                                    listener.deleteRouteBtnClicked(getItem(position))
                                    true
                                }
                                else -> false
                            }
                        }
                        popupMenu.show()
                    }
                }
            }
        }

        fun bind(route: Route){
            binding.routeNameTv.text = route.name

            if (route.selected) {
                binding.apply {
                    container.setBackgroundResource(R.drawable.navigation_view_route_selected_bg)
                    routeNameTv.setTextColor(Color.parseColor("#167ef9"))
                    routeNameTv.typeface = Typeface.DEFAULT_BOLD
                    routeMenuIb.setImageResource(R.drawable.ic_route_menu_blue)
                    routeIcIv.setImageResource(R.drawable.ic_route_blue)
                }
            } else {
                binding.apply {
                    container.setBackgroundResource(R.drawable.navigation_view_route_bg)
                    routeNameTv.setTextColor(Color.parseColor("#515767"))
                    routeNameTv.typeface = Typeface.DEFAULT
                    routeMenuIb.setImageResource(R.drawable.ic_route_menu_black)
                    routeIcIv.setImageResource(R.drawable.ic_route_black)
                }
            }
        }
    }

    class DiffCallback : DiffUtil.ItemCallback<Route>(){

        override fun areItemsTheSame(oldItem: Route, newItem: Route) =
                oldItem.routeId == newItem.routeId

        override fun areContentsTheSame(oldItem: Route, newItem: Route) =
                oldItem == newItem
    }
}